android/src/main/java/com/guichaguri/trackplayer/module/MusicModule.kt

```diff
@@ -185,21 +185,42 @@ class MusicModule(private val reactContext: ReactApplicationContext?) :
     fun add(data: ReadableArray?, insertBeforeIndex: Int, callback: Promise) {
         if (verifyServiceBoundOrReject(callback)) return
 
+        val tracks: MutableList<Track> = mutableListOf()
         val bundleList = Arguments.toList(data)
-        val tracks: List<Track> = try {
-            Track.createTracks(
-                reactApplicationContext, bundleList, RatingCompat.RATING_HEART
-            )!!
-
-
-        } catch (ex: Exception) {
-            callback.reject("invalid_track_object", ex)
-            return
+        if (bundleList != null) {
+            for (trackBundle in bundleList) {
+                if (trackBundle is Bundle) {
+                    try {
+                        tracks.add(
+                            Track(reactApplicationContext, trackBundle, RatingCompat.RATING_HEART)
+                        )
+                    } catch(ex: Exception) {
+                        callback.reject("invalid_track_object", ex)
+                    }
+                } else {
+                    callback.reject("invalid_track_object", "Track was not a dictionary type")
+                }
+            }
+        } else {
+            callback.reject("invalid_parameter", "Was not given an array of tracks")
         }
 
-        musicService.apply {
-            add(tracks)
-            callback.resolve(null)
+        when {
+            insertBeforeIndex < -1 || insertBeforeIndex > musicService.tracks.size -> {
+                callback.reject("index_out_of_bounds", "The track index is out of bounds")
+            }
+            insertBeforeIndex == -1 -> {
+                musicService.apply {
+                    add(tracks)
+                    callback.resolve(null)
+                }
+            }
+            else -> {
+                musicService.apply {
+                    add(tracks, insertBeforeIndex)
+                    callback.resolve(null)
+                }
+            }
```

Comment:
`If else` might be more suitable here. Up to you.

Reply:
The kotlin compiler suggested I use when instead of if's :)

Reply:
I usually consider `when`s more like a `switch`. And in this case where we have different types of conditioning I prefer to use `if else`. But you can leave it like this if you prefer.

Reply:
I do what the compiler tells me to do :) Since its about code style and we haven't yet decided on a type let's keep it like this.